core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1597,6 +1640,327 @@ public void addStateChangeListener(StateChangeListener<DistributedStagesSchedule
+                List<ListenableFuture<Void>> blockedStages = new ArrayList<>();
+                while (!isFinishingOrDone(queryStateMachine) && !stateMachine.getState().isDone()) {
+                    blockedStages.clear();
+                    boolean atLeastOneStageIsNotBlocked = false;
+                    boolean allFinished = true;
+                    for (FaultTolerantStageScheduler scheduler : schedulers) {
+                        if (scheduler.isFinished()) {
+                            continue;
+                        }
+                        allFinished = false;
+                        ListenableFuture<Void> blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                            continue;
+                        }
+                        try {
+                            scheduler.schedule();
+                        }
+                        catch (Throwable t) {
+                            stateMachine.transitionToFailed(t, Optional.of(scheduler.getStageId()));
+                            return;
+                        }
+                        blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                        }
+                        else {
+                            atLeastOneStageIsNotBlocked = true;
+                        }
+                    }
+                    if (allFinished) {
+                        stateMachine.transitionToFinished();
+                        return;
+                    }
+                    // wait for a state change and then schedule again
+                    if (!atLeastOneStageIsNotBlocked && !blockedStages.isEmpty()) {
+                        try (TimeStat.BlockTimer timer = schedulerStats.getSleepTime().time()) {
+                            try {
+                                tryGetFutureValue(whenAnyComplete(blockedStages), 1, SECONDS);
+                            }
+                            catch (CancellationException e) {
+                                log.debug("Future cancelled");
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Throwable t) {
+                stateMachine.transitionToFailed(t, Optional.empty());
+                schedulers.forEach(FaultTolerantStageScheduler::abort);
+                closeNodeAllocator();
+            }
+        }
+
+        private static boolean isFinishingOrDone(QueryStateMachine queryStateMachine)
+        {
+            QueryState queryState = queryStateMachine.getQueryState();
+            return queryState == FINISHING || queryState.isDone();
+        }
+
+        @Override
+        public void cancelStage(StageId stageId)
+        {
+            // single stage cancellation is not supported in fault tolerant mode
```

Comment:
`throw new UnsupportedOperationException();`?

QQ: when would someone want to cancel a stage? Isn't that effectively cancelling the whole query>

Reply:
Changed to `UnsupportedOperationException`. It is possible to cancel a single stage via the REST API (e.g.: for testing / debugging)